examples/avro-cli.py

```diff
@@ -152,21 +176,21 @@ def consume(topic, conf):
 
 
 def main(args):
-    # handle common configs
-    conf = {'bootstrap.servers': args.bootstrap_servers,
-            'schema.registry.url': args.schema_registry}
-
+    registry_client = SchemaRegistryClient(url=args.schema_registry)
     if args.userinfo:
-        conf['schema.registry.basic.auth.credentials.source'] = 'USER_INFO'
-        conf['schema.registry.basic.auth.user.info'] = args.userinfo
+        registry_client.user_auth(args.userinfo.split(':'))
+
+    avro_serializer = AvroSerializer(registry_client, record_schema,
```

Comment:
the approach of passing the schema and converter to the constructor seems defensible, but...

My feeling is still to err on the side of simplicity. Primarily, I'm struggling to see the value of the Converter idea vs just requiring you need to pass in a "dict convertible like object" (i.e. one with a `to_dict()` method) into the produce method otherwise it blows up. The latter seems easier to grok (one less abstraction), and pythonic. I'd would be convinced to change my opinion on this if there's precedent from other well used python libraries (i don't know, don't use them).

An argument against this is if people typically don't want to/can't have a `to_dict` method on their types they want to produce. I feel most of the time this won't be a problem and if it is, it's a simple matter to provide the behavior in a sub class. Also, if it turns out people do do the subclass thing all the time, the `Converter` concept can be added later. On the flip side of this, if we add it now, we can't take it away. Err on the side of smallest API is a good strategy. 

Second, I was wondering if even the record_schema could be specified via a `.schema()` method on the object passed into the produce method. That I think is not a good idea though for reasons of performance, and also what most closely matches the behavior in other languages is to specify it in the constructor (in other languages it's a static property of the type... but close).

Reply:
I think it makes more sense to modify the behavior of the serializer, through the use of the converter, than it does to modify the behavior of the object being serialized. The object may be serialized as Avro today, Protobuf tomorrow and JSON the next. It could become quite messy if I had to add methods to satisfy the needs of each serializer.

Reply:
i remain unconvinced and we can argue further.... widespread precedent of use of abstract base classes in python would help the case against what I think here. but it seems to go against what I thought the language was about.

Reply:
The JSON library takes a very similar approach with its custom encoder and decoder functions which allow you to customize how your object will go from its initial form to the json representation that comes out. We actually take advantage of this in the SchemaRegistryClient code.

Reply:
I strongly object to require modifications or sub-classing of the original object.
Having said that, I'm not that fond of Converts since it adds another level of abstraction and introduces a new concept.
Have we considered sub-classing the Serializer type instead, so:
```python
class UserAvroSerializer(AvroSerializer):
    def __init__(kwargs):
       super(UserAvroSerializer, self).__init__(**kwargs)
    def to_dict(user):
       return { 'id': user.id, 'name': user.name, 'breadth': user.breadth }
    def from_dict(data):
       return User(**data)
```